tokio/src/sync/broadcast.rs

```diff
@@ -0,0 +1,980 @@
+    ///
+    /// ```
+    /// use tokio::sync::broadcast;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, _rx1) = broadcast::channel(16);
+    ///
+    ///     assert_eq!(1, tx.receiver_count());
+    ///
+    ///     let mut _rx2 = tx.subscribe();
+    ///
+    ///     assert_eq!(2, tx.receiver_count());
+    ///
+    ///     tx.send(10).unwrap();
+    /// }
+    /// ```
+    pub fn receiver_count(&self) -> usize {
+        let tail = self.shared.tail.lock().unwrap();
+        tail.rx_cnt
+    }
+
+    fn send2(&self, value: Option<T>) -> Result<(), SendError<Option<T>>> {
+        let mut tail = self.shared.tail.lock().unwrap();
+
+        if tail.rx_cnt == 0 {
+            return Err(SendError(value));
+        }
+
+        // Position to write into
+        let pos = tail.pos;
+        let rem = tail.rx_cnt;
+        let idx = (pos & self.shared.mask as u64) as usize;
+
+        // Update the tail position
+        tail.pos = tail.pos.wrapping_add(1);
+
+        // Get the slot
+        let slot = &self.shared.buffer[idx];
+
+        // Acquire the write lock
+        let mut prev = slot.lock.fetch_add(1, SeqCst);
+
+        while prev & !1 != 0 {
+            // Concurrent readers, we must go to sleep
+            tail = self.shared.condvar.wait(tail).unwrap();
+
+            prev = slot.lock.load(SeqCst);
+        }
+
+        // Release the mutex
+        drop(tail);
+
+        // Slot lock acquired
+        slot.write.pos.with_mut(|ptr| unsafe { *ptr = pos });
+        slot.write.val.with_mut(|ptr| unsafe { *ptr = value });
+
+        // Set remaining receivers
+        slot.rem.store(rem, SeqCst);
+
+        // Release the slot lock
+        slot.lock.store(0, SeqCst);
+
+        // Notify waiting receivers
+        self.notify_rx();
+
+        Ok(())
+    }
+
+    fn notify_rx(&self) {
+        let mut curr = self.shared.wait_stack.swap(ptr::null_mut(), SeqCst) as *const WaitNode;
+
+        while !curr.is_null() {
+            let waiter = unsafe { Arc::from_raw(curr) };
+
+            // Update `curr` before toggling `queued` and waking
+            curr = waiter.next.with(|ptr| unsafe { *ptr });
+
+            // Unset queued
+            waiter.queued.store(false, SeqCst);
+
+            // Wake
+            waiter.waker.wake();
+        }
+    }
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Sender<T> {
+        let shared = self.shared.clone();
+        shared.num_tx.fetch_add(1, SeqCst);
+
+        Sender { shared }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        if 1 == self.shared.num_tx.fetch_sub(1, SeqCst) {
+            let _ = self.send2(None);
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Lock the next value if there is one.
+    ///
+    /// The caller is responsible for unlocking
+    fn recv_ref(&mut self, spin: bool) -> Result<RecvGuard<'_, T>, TryRecvError> {
+        let idx = (self.next & self.shared.mask as u64) as usize;
```

Comment:
nit/tioli: should `Shared` maybe have a method like
```rust
fn index(&self, i: u64) -> usize {
    (i & self.shared.mask as u64) as usize;
}
```
it seems like this is repeated in a few places